LEG-45 - Support for Impact Assessments with enhancements

QR-TNA opened the pull request:
# Impact Assessment Support

This pull request introduces comprehensive support for Impact Assessment (IA) documents in the legislation parser, transforming them into semantically enriched Akoma Ntoso (AKN) structures that comply with the Associated Documents in XML specification, without altering their visual presentation.

The implementation includes foundational refactoring work to generalise the parser beyond Explanatory Memoranda (EMs), making the codebase modular and extensible for other document types.

## Summary of Changes

### Phase 0: General Refactor
- Refactored `/src/leg/` to remove EM-specific assumptions
- Introduced a modular architecture to support multiple document types (e.g. IA, EM)
- Extracted and centralised shared functionality into common helpers and interfaces
- Added `BaseLegislativeDocumentParser` and `BaseHelper` for code reuse

### Phase 1: IA Parsing Framework
- Added IA parsing framework under `/src/leg/ia/`
- Implemented IA-specific models and helper classes
- Added IA-specific XSD schema (`ia-subschema.xsd`) with custom inline elements
- Integrated IA support into main command-line interface (`--hint ia`)

### Phase 2: Semantic Transformation

#### Inline Metadata Extraction
- Introduced semantic element mappings for IA-specific metadata:
  - Document title, stage, date
  - Lead department/agency (using custom `<inline name="leadDepartment">`)
  - Other departments (using custom `<inline name="otherDepartments">`)
- Centralised metadata mapping in `Helper.cs`
- Updated `Builder.cs` to construct IA-specific inline AKN elements

#### Header Structure Enhancement
- Added semantic containers: `<hcontainer name="summary">`
- Applied consistent `eId` attributes and unified header layout
- Fixed layout issues including floating sections and empty rows
- Transformed header from `<level>` to `<section>` for proper structure

#### Section-Based Organisation
- Detected and organised key IA sections:
  - `cost-analysis`, `policy-review`, `policy-objectives`, `policy-options`
  - `economic-assessment`, `business-assessment`, `equalities`, `sustainability`
- Wrapped each section in a semantic `<section>` element with:
  - Stable `eId` identifiers
  - Hidden semantic headings for machine readability

### Phase 3: URI Generation & Metadata Integration

#### Spec-Compliant URI Scheme
- Implemented URI generation following the Associated Documents specification:
  - Format: `{legislation-type}/{year}/{number}/impact-assessment/{stage}`
  - Example: `uksi/2018/1149/impact-assessment/final`
- Added stage normalization function mapping CSV values to URL-safe format
- Falls back to `ukia/{year}/{number}/impact-assessment/{stage}` when no legislation mapping exists
- Stage component gracefully omitted if not available

#### IA to Legislation Mapping
- Implemented CSV-based mapping system (`ia_to_legislation_mapping.csv`)
- Maps UKIA identifiers (year/number) to legislation URIs
- Extracts full metadata records including:
  - Document stage, main type, department
  - IA date, PDF date, modified date
  - Legislation class, year, number
  - Title and other metadata fields

#### Metadata in Proprietary Section
- Added comprehensive metadata to AKN `<proprietary>` section using `uk:` namespace:
  - `<uk:documentStage>` - Document stage (Final, Enactment, etc.)
  - `<uk:documentMainType>` - Document main type
  - `<uk:department>` - Lead department
  - `<uk:iaDate>`, `<uk:pdfDate>` - Important dates
  - `<uk:legislationClass>` - Associated legislation class
- Metadata structure matches original XML metadata files

### Phase 4: Schema Compliance & Validation
- Extended `ia-subschema.xsd` to support custom inline elements with `name` attributes
- Updated schema to allow inline elements in header paragraphs and content
- Integrated validation into test suite (all IA tests validate against schema)
- All generated AKN documents pass XSD validation

## Testing

Parse IA document with semantic enrichment:
`dotnet run -- --input test/leg/ia/original\ filenames/ukia_20250001_en.docx --hint ia --output test1.akn`

Run IA-specific test suite (32 tests, all passing):
`dotnet test test/test.csproj --filter "FullyQualifiedName~UK.Gov.Legislation.ImpactAssessments"`

Validate schema compliance:
`dotnet test test/test.csproj --filter "FullyQualifiedName~UK.Gov.Legislation.ImpactAssessments.Test.TestIA.ValidateParsedOutput"`

### Core Implementation
- `src/leg/ia/Helper.cs` - Main transformation logic with semantic mappings
- `src/leg/ia/Parser.cs` - IA-specific parser implementation
- `src/leg/ia/IAMetadata.cs` - Metadata handling and URI generation
- `src/leg/common/IALegislationMapping.cs` - CSV mapping and URI building
- `src/leg/common/IAMappingRecord.cs` - Data model for CSV records
- `src/leg/Builder.cs` - AKN document construction with proprietary metadata
- `src/Program.cs` - Command-line interface integration

### Schema & Configuration
- `src/leg/schemas/ia-subschema.xsd` - IA-specific AKN schema with inline element support
- `src/leg/source/ia_to_legislation_mapping.csv` - IA to legislation mapping data

### Tests
- `test/leg/TestIA.cs` - Comprehensive test suite with validation
- `test/leg/ia/original filenames/*.akn` - Regenerated test files with new URI structure

## Backwards Compatibility

All changes are fully backwards compatible:
- Existing EM parsing remains unchanged
- No breaking changes to public APIs
- All existing tests continue to pass

## Schema Compliance

- All generated AKN documents validate against `ia-subschema.xsd`
- Custom inline elements properly defined in schema
- FRBR metadata structure follows AKN 3.0 specification
- Proprietary metadata uses standard `uk:` namespace

QR-TNA committed:
feat(ia): scaffold Impact Assessments transformation from Explanatory Memoranda

- Copied core transformation files from 'em' (Explanatory Memoranda) to 'ia' (Impact Assessments)
- Renamed all namespaces, class names, and references from ExplanatoryMemoranda to ImpactAssessments

This provides a foundation for further development of Impact Assessments transformation, based on the proven Explanatory Memoranda implementation.

QR-TNA committed:
feat: add validation and error handling for invalid hint values

QR-TNA committed:
feat: add support for 'ia' (Impact Assessment) document transformation

- Added 'ia' as a valid value for the --hint option.
- Integrated IA transformation logic using UK.Gov.Legislation.ImpactAssessments.
- Implemented TransformIA method, mirroring EM handling.
- Updated option description and validation to include 'ia'.

QR-TNA committed:
add IA samples to the test folder

QR-TNA committed:
feat(IA): change references to EM in IA files

QR-TNA committed:
feat(IA): add IA template to repo

QR-TNA committed:
feat(IA): change IA template to docx

QR-TNA committed:
feat(IA): delete old IA template

QR-TNA committed:
Merge branch 'main' into add-support-for-impact-assessments

QR-TNA committed:
feat: Add Impact Assessment support with dedicated schema and CSS class mapping

- Create ia-subschema.xsd with IA-specific CSS classes
- Implement content-based detection for IA styling (ia-title, ia-head-label, etc.)
- Fix document type to use "ImpactAssessment"
- Enable complete IA DOCX → AKN XML → HTML pipeline

QR-TNA committed:
docs: Add README for legislation module (src/leg/)

Documents the EM and IA parsers architecture, usage patterns, and provides guidance for adding new document types.

QR-TNA committed:
refactor: Reorganise legislation schemas into schemas/ folder and rename for consistency

- Move subschema.xsd → schemas/em-subschema.xsd
- Move ia-subschema.xsd → schemas/ia-subschema.xsd
- Update project file and Validator.cs to use new schema paths
- Update README to reflect new schema organisation

QR-TNA committed:
feat: Enhance IA styling with Arial font, solid borders, and improved content detection

- Apply Arial font to all text in Impact Assessment documents
- Change table borders from dotted to solid for IA documents
- Add black background with white text for ia-title elements
- Simplify content pattern matching by removing HTML tags first
- Remove redundant font-family declarations and debug code
- Clean up unnecessary Structure.cs changes that had no effect

QR-TNA committed:
fix: Add support for cx1 (Chart Extensions 1.0) in AlternateContent parsing

Handle Microsoft Office 2016 Chart Extensions by using fallback content when encountering cx1 alternate content blocks. This resolves parsing failures for documents containing modern chart elements.

QR-TNA committed:
feat: Add Impact Assessment (IA) test framework support

- Add embedded resource entries for IA test files in test.csproj
- Fix test discovery to use embedded resources instead of file system
- Add TestIA.cs with proper resource-based test data discovery
- Enable 9 IA tests (3 passing, 6 pending expected XML files)

Fixes test framework No data found error and enables proper IA parsing validation.

QR-TNA committed:
refactor: remove unnecessary schema validation from IA processing

QR-TNA committed:
test: skip IA tests when corresponding XML file is missing

QR-TNA committed:
Merge branch 'main' into add-support-for-impact-assessments

QR-TNA committed:
fix: revert Builder.cs as changes were breaking TestEM

QR-TNA committed:
feat(ia-schema): enhance legal citation support with ref element and improved links

QR-TNA committed:
refactor: eliminate code duplication in legislation parsers

- Extract common parsing logic into base classes (BaseLegislativeDocumentParser,
  BaseHeaderSplitter, BaseMetadata, BaseHelper)
- Introduce LegislativeDocumentConfig for document-specific settings
- Move identical RegulationNumber class to common
- Organize models into models/ directory with unified namespace
- Tests still pass

QR-TNA committed:
chore: remove old xml output files

QR-TNA committed:
style: make improvements to css for IA tables

QR-TNA committed:
style: added grey styling to some cells to copy the docx style

QR-TNA committed:
refactor: remove dead code from xsl and try fix for class names showing in footnotes

QR-TNA committed:
fix: table indentation in Impact Assessment documents

QR-TNA committed:
feat: standardise header text styling with ia-header-text class

QR-TNA committed:
refactor: optimize Impact Assessment styles in XSL transform

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
fix(LEG-38): resolve ia-subschema.xsd validation errors

QR-TNA committed:
refactor: simplify and clean up Impact Assessment CSS in akn2html.xsl

QR-TNA committed:
fix: preserve spaces in bold/italic/underline elements in akn2html.xsl

QR-TNA committed:
fix: remove overly broad CSS selectors causing incorrect grey cell backgrounds

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
test: update TestIA to use .akn files instead of .xml

QR-TNA committed:
feat: Enhanced parsing diagnostics and console logging

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
feat: add semantic inline elements for IA metadata

- Add docTitle, docStage, docDate, docDepartment inline elements to models
- Update Builder.cs to handle new semantic inline elements in AKN output
- Enhance IA Helper to transform metadata labels into semantic elements
- Update IA schema to support new inline elements in paragraphs
- Clean up akn2html.xsl: remove court-specific elements, fix namespace

QR-TNA committed:
feat: implement semantic inline elements and refactor parsing logic

QR-TNA committed:
feat: implement comprehensive semantic structure improvements

QR-TNA committed:
fix: resolve failing test suite issues

- fix parser version detection by enabling GenerateAssemblyInfo in judgments.csproj
- fix Explanatory Memoranda URL format from '/em' to '/memorandum' suffix
- add missing AKN reference files for Impact Assessment tests

QR-TNA committed:
fix: place semantic tags around values instead of labels

- Update TransformToSemanticElement method to wrap semantic tags around content values rather than label text
- Change markup from <b><docTitle>Title</docTitle></b>: content to <b>Title:</b> <docTitle>content</docTitle>
- Add date attribute parsing for docDate elements (e.g., date="2015-09-30")
- Remove obsolete CreateBoldSemanticElement method
- Update test expectations to match new semantic structure

Fixes semantic markup structure to properly identify document metadata values in IA documents.

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
feat: remove name attribute and update test files

QR-TNA committed:
fix: ensure Impact Assessment documents validate against AKN 3.0 schema

Modified AKN builder to skip elements not supported in strict AKN 3.0 for
doc elements (Impact Assessments):
- Skip images (img), math (subFlow), tabs in anchors (marker), and TOC
- Replace th with td elements in tables
- Generate inline numbered paragraphs instead of blockContainer in tables
- Skip attachments for doc elements per IA subschema

Modified IA-specific processing:
- Remove docNumber semantic element (not allowed in <p>)
- Remove date attribute from docDate (not declared in schema)
- Clean up empty heading elements that violate sequence constraints

Added test infrastructure:
- ValidateAkn test method to validate all IA test files against schema
- RegenerateAllTestFiles utility method for test maintenance
- Regenerated all 9 IA test .akn files with schema-compliant output

QR-TNA committed:
test: add validation tests and regenerate test files

QR-TNA committed:
feat: support annexes in Impact Assessment and Explanatory Memoranda documents

Extended the IA subschema to allow attachments/annexes while maintaining
Akoma Ntoso 3.0 compliance. Previously, annexes were being parsed but
silently discarded during XML serialisation.

Changes:
- Extended ia-subschema.xsd to allow <attachments> element in doc
- Added "annex" to allowed doc @name attribute values
- Extended block-container type to allow <blockContainer> elements
- Restored AddAnnexes() method in Builder.cs
- Updated AddAnnex() to wrap content in hcontainer > content structure
- Regenerated test files that have annexes: test/leg/ia/test13.akn, test/leg/em/test4.akn

Test results: 253 passed, 0 failed

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
fix: ensure AKN 3.0 compliance for docProponent and docDate

- Replace invalid `docDepartment` with `docProponent`
- Add required `date` attribute to `docDate` elements
- Implement date parsing for UK dates and various formats
- Update schemas and regenerate all IA test files

QR-TNA committed:
fix: Improve RegenerateAkn.cs path resolution

QR-TNA committed:
fix: use document date in FRBR metadata and move lastModified to proprietary

BREAKING CHANGE: IA FRBR dates now use document date; EM structure unchanged

Impact Assessments:
- Replace `docDepartment` with AKN-compliant `docProponent`
- Add required `date` attribute to `docDate` elements
- FRBRWork/FRBRExpression now use document date with name="document"
- Move lastModified to proprietary section as `<dc:modified>`
- Declare dc namespace at doc level for cleaner XML structure

Explanatory Memoranda:
- Continue using lastModified in FRBRExpression (appropriate for EMs)
- Update regeneration test for better CI compatibility

Regenerate all IA and EM test files with corrected metadata structure.

QR-TNA committed:
chore: add dc:modified to proprietary for all EM and IA test files

Regenerate all legislative document test files to include dc:modified
in proprietary section for consistency across document types.

jurisdatum commented on src/leg/Builder.cs:
```diff
@@ -190,10 +209,21 @@ private void AddAnnex(XmlElement attachments, IAnnex annex, int n, DocumentMetad
+        // Now that we've extended the schema to allow blockContainer in block-container type,
+        // we can use the simpler content structure
+        XmlElement hcontainer = doc.CreateElement("hcontainer", ns);
+        hcontainer.SetAttribute("name", "annexContent");
```
I'm not wild about the idea of wrapping every annex's contents in an hcontainer. What's the reason for it? Just to simplify the subschema?

jurisdatum commented on src/leg/schemas/em-subschema.xsd:
You should add `<xs:element ref="attachments" minOccurs="0" />` to the definition of `<xs:element name="doc">`. Test4 currently fails validation.

QR-TNA committed:
fix: add attachments element to EM subschema to support annexes

QR-TNA committed:
fix: prevent creation of invalid docDate elements without date attribute

QR-TNA committed:
refactor: remove unnecessary hcontainer wrapper from annexes

Remove the hcontainer wrapper around annex content by allowing block
elements (p, table, blockContainer) directly in mainBody for both EM
and IA schemas. This aligns with Akoma Ntoso 3.0 standard which allows
both hierarchical and block elements in mainBody.

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
Merge branch 'LEG-45' into LEG-77

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
Merge branch 'main' into LEG-45

jurisdatum reviewed (changes requested):
Can we drop the `RegenerateAkn` class? I believe `TestEM.RegenerateAllTestFiles()` and `TestIA.RegenerateAllTestFiles()` already cover the same regeneration flow when we need it. If there’s a reason to keep this class, could we mark those two `[Fact]`s as skipped (`[Fact(Skip = "Manual regeneration helper")]`) so `dotnet test` doesn’t start rewriting fixtures every run? Keeps CI happy while still letting us run the helpers on demand.

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
refactor(test): remove duplicate RegenerateAkn class

QR-TNA committed:
Merge branch 'LEG-45' of https://github.com/nationalarchives/tna-judgments-parser into LEG-45

jurisdatum commented on src/leg/ia/Helper.cs:
```diff
@@ -0,0 +1,518 @@
+        // { "IA No:", "docNumber" },  // Commented out: not valid in AKN 3.0 schema for <p> elements
+        { "Stage:", "docStage" },
+        { "Date:", "docDate" },
+        { "Lead department or agency:", "docProponent" },
```
I'm not sure `<docProponent>` is the best element for this purpose. I always assumed it was for things like the sponsor of a bill, but I'm not sure. Let's discuss. You can always use a custom `<inline name="...">` element.

jurisdatum commented on src/leg/ia/Helper.cs:
```diff
@@ -0,0 +1,518 @@
+            string cleanContent = CleanContent(p.InnerText?.Trim() ?? "");
+            
+            // Try to transform to semantic element first
+            bool transformed = TryTransformToSemanticElement(xml, p, cleanContent);
```
You might want to think about expanding the semantic recognition code so it can handle things that span a line break. So, for example, the single-line-only approach leads to this, in test13
```xml
<p><b>Lead department or agency:</b> <docProponent>Department for Business,</docProponent></p>
<p class="ia-header-text">Energy and Industrial Strategy             </p>
```
In the judgments parser, I do this sort of enrichment to the model before building the XML, which I find easier. Also, it makes the semantic recognition independent of the target XML format.

QR-TNA committed:
feat: add IA metadata from filename and legislation mapping

- Add ia_to_legislation_mapping.csv as embedded resource for IA to legislation URI lookup
- Create IALegislationMapping helper to parse filenames (ukia_YYYYNNNN_en) and lookup legislation URIs
- Create IAMetadata class for IA-specific metadata generation
- Add LegislationUri property to DocumentMetadata
- Update IA Helper.Parse() to accept filename parameter
- Update Builder to output <uk:legislation> element in proprietary section
- Update CLI to pass filename to IA parser
- Update IA tests to use new filename-based test files
- Add inline element support to IA subschema

The IA AKN output now includes:
- Complete FRBR URIs derived from filename (e.g., ukia/2025/1)
- Legislation reference in <uk:legislation> from CSV mapping

QR-TNA committed:
fix(ia): fix IA schema validation issues

- Move tables from preface to mainBody (preface only allows p elements)
- Fix duplicate/mispositioned headings in sections (keep first, remove extras)
- Fix nested anchor elements by unwrapping inner anchors

QR-TNA committed:
fix: handle edge cases in REF field and inline parsing with content preservation

- REF field: when target bookmark has no number, use paragraph text content
  instead of throwing exception (preserves content that would be lost)
- REF field: handle case where bookmark has no parent paragraph gracefully
- Inline parsing: recursively parse children of unknown element types
  (following the same pattern as InsertedRun, SdtRun, smartTag, etc.)
- Inline parsing: fallback to InnerText if element has no parseable children
- Only log warning if truly no content can be extracted

These fixes enable parsing of ukia_20250003_en and ukia_20250010_en
which previously crashed, while preserving document content.

QR-TNA committed:
Merge LEG-77 into LEG-45: resolve IA parser conflict

QR-TNA committed:
Fix merge issues: add LastModified property, restore MovePrefaceTablesToMainBody call, regenerate test files

QR-TNA committed:
IA parser improvements: semantic elements, section structure, schema updates

QR-TNA committed:
Update IA URI format to spec: {type}/{year}/{number}/impact-assessment

QR-TNA committed:
feat: extract full metadata from CSV mapping for Impact Assessments

Add comprehensive metadata extraction from CSV mapping file (derived from XML metadata)
and include in AKN proprietary section using uk: namespace elements.

- Add IAMappingRecord class to hold complete CSV metadata structure
- Extend IALegislationMapping to parse full CSV records with quoted field handling
- Add metadata properties to IAMetadata: DocumentStage, DocumentMainType, Department, IADate, PDFDate, LegislationClass
- Populate metadata from CSV lookup by year/number from filename
- Add all metadata fields to AKN proprietary section in Builder.cs
- Fix CSV parsing bug: legislation URI was using wrong column index (title instead of legislation_uri)

Metadata fields now included in AKN output:
- uk:documentStage (Final, Enactment, etc.)
- uk:documentMainType (UnitedKingdomImpactAssessment, etc.)
- uk:department (Lead department)
- uk:iaDate, uk:pdfDate (dates)
- uk:legislationClass (when legislation exists)

QR-TNA committed:
feat: add stage component to impact assessment URIs

- Add NormalizeStage() function to map CSV stage values to URL-safe format
- Update BuildShortUriComponent() to include stage in URI path
- Update IAMetadata to pass document stage to URI builder
- Update test XSLT to exclude new metadata fields from comparison
- Regenerate all test files with new URI structure including stage

URIs now follow format: {type}/{year}/{number}/impact-assessment/{stage}
Example: uksi/2018/1149/impact-assessment/final

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
fix: extract dropdown values from FORMDROPDOWN fields

Handle both explicit selections and default values:
- When result element exists in ddList, use its 0-based index to get selected value
- When no result element exists, use first item as default (Word displays default but doesn't save selection)

Fixes extraction of dropdown values like 'Source of intervention' and 'Type of measure' in Impact Assessment documents.

QR-TNA committed:
feat: add SemanticEnricher for model-level semantic element recognition

- Add SemanticEnricher class to recognise semantic patterns in Impact Assessments
- Recognizes patterns like 'Title:', 'Stage:', 'Date:', 'Lead department or agency:', etc.
- Handles multi-line patterns where semantic content spans multiple blocks
- Creates semantic inline elements (DocTitle, DocStage, DocDate, LeadDepartment, OtherDepartments)
- Integrates with IA parser to enrich header and body blocks
- Regenerate test .akn files with semantic enrichment

QR-TNA committed:
fix: remove invalid table class and tbody elements from AKN output

- Remove invalid 'ia-table' class attribute from table elements
  (not in schema enumeration: ia-heading, ia-box, ia-costs, ia-benefits, ia-policy-options)
- Remove invalid tbody wrapper elements from tables
  (AKN schema requires tr elements directly as children of table)
- Fix table structure in GenerateTableOfContents method
- Regenerate all test .akn files with valid structure

QR-TNA commented on src/leg/ia/Helper.cs:
```diff
@@ -0,0 +1,518 @@
+        var logger = Logging.Factory.CreateLogger<Helper>();
+
+        // Define major IA sections with their identifying content patterns
+        var sectionMappings = new Dictionary<string, string> {
```
Unused at the moment so removed.

QR-TNA committed:
refactor(ia): simplify section pattern matching to use array instead of dictionary

QR-TNA committed:
refactor: improve code clarity and remove blockContainer post-processing

- Remove inconsistent phase numbering from method comments
- Remove dead code in DetermineCssClass (unreachable inTable check)
- Simplify collection creation in FixNestedAnchors using LINQ
- Move blockContainer handling from Helper to Builder to prevent creation in table cells
- Remove ConvertBlockContainerInTables method (no longer needed)
- Update test files to reflect structural changes

QR-TNA committed:
feat: extend schema to support images and math elements

- Add img element support to ia-subschema.xsd with src, width, height, style, alt attributes
- Add subFlow and foreign elements for mathematical formulas
- Remove code that was stripping img and subFlow elements from Helper.cs
- Regenerate test files to reflect schema changes

QR-TNA committed:
feat: change language code from 'eng' to 'en' to match CLML convention

Update FRBRlanguage attribute to use ISO 639-1 two-letter code 'en' instead of three-letter 'eng' in leg folder documents. This aligns with the CLML convention where 'en' is used for English and 'cy' will be used for Welsh.

- Updated Builder.cs to use 'en' for language attribute
- Regenerated all EM and IA test files with new language code

QR-TNA committed:
fix: change all date formats from xs:dateTime to xs:date

Update all date attributes to use xs:date format instead of xs:dateTime to meet MarkLogic requirements:
- dc:modified now outputs date only (e.g., 2024-11-28)
- FRBRManifestation/FRBRdate with name='transform' now outputs date only
- All other dates already used date format

This aligns with MarkLogic expectations for associated documents.

QR-TNA committed:
feat: add uk:IA element and fix date parsing

- Add uk:IA element with full UKIA URI to proprietary section
- Add uk:ukiaYear, uk:ukiaNumber, uk:legislationYear, uk:legislationNumber
  elements for easier MarkLogic loading
- Fix date parsing to handle trailing asterisks (e.g., "13/12/2012*")
- Fix date parsing to handle parenthetical notes (e.g., "May 2017 (updated...)")
- Fix legislation_number parsing to remove .0 suffix from CSV numbers

QR-TNA committed:
feat: add section numbers to TOC headings

Include section numbers in table of contents headings for better navigation
on the website contents page. Format: "2. Cost of Preferred (or more likely) Option"

Changes:
- Modified TOC generation to prepend section number to heading text
- Updated all IA test files with new TOC format

QR-TNA committed:
feat(ia): update URI path to /impacts/ and add whole document TOC link

- Change IA URI path segment from 'impact-assessment' to 'impacts'
- Add unnumbered 'The whole Impact Assessment' link at top of TOC
- Regenerate all IA test files with updated URIs and TOC structure

QR-TNA committed:
fix(ia): use UKIA year/number in URI path instead of stage

- Update URI format from /impacts/{stage} to /impacts/{ukia-year}/{ukia-number}
- Align with live website pattern (e.g., /ukdsi/2010/.../impacts/2009/207)
- Stage remains in metadata (uk:documentStage) but not in URI path
- Supports multiple IAs at same stage for same legislation
- Regenerate all IA test files with corrected URI structure

QR-TNA committed:
feat(ia): implement hierarchical TOC structure with guaranteed whole document link

- Set 'The whole Impact Assessment' to level 1
- Set all section entries to level 2
- Ensure TOC is always generated even when no sections exist
- Regenerate all IA test files with updated TOC hierarchy

QR-TNA committed:
feat(em): implement versioned URI generation with CSV mapping

Implements EM URI generation following the same pattern as IAs, with URIs in the format {type}/{year}/{number}/memoranda/{version}. Version numbering is determined by CSV mapping data, handling both legacy filenames and post-2023 _NNN suffix convention.

- Add EMLegislationMapping for CSV-based metadata lookup
- Add EMMetadata class with version-based URI generation
- Support dynamic path segments (memoranda, executive-note, policy-note)
- Add full CSV metadata to AKN proprietary section
- Update test infrastructure to use original filenames pattern
- Regenerate all 22 EM test files with correct URIs

QR-TNA committed:
feat(em): add table of contents support for Explanatory Memoranda

Implements TOC generation for Explanatory Memoranda following the same pattern as Impact Assessments. The TOC includes a "whole document" link and sequential numbering of all sections with headings.

- Add GenerateTableOfContents method to EM Helper
- Support both paragraph/intro and section/heading EM structures
- Generate sequential TOC numbering starting from 1
- Add eId attributes to paragraphs and sections for proper linking
- Document type-specific labels (Explanatory Memorandum, Policy Note, Executive Note)
- Insert TOC at beginning of mainBody
- Regenerate all 22 EM test files with TOC

QR-TNA committed:
feat(leg): detect IA sections by structure and fix summary TOC label

- Replace content-based TransformContentSections with structural pattern
  detection: bold numbered/lettered paragraphs and levels with bold headings
- Prefer hcontainer name attribute (e.g. Summary) over content extraction
  for TOC heading
- Regenerate IA test AKN fixtures

QR-TNA committed:
feat(ia): group section content into intro/hier/wrapUp and regenerate tests

Use AKN structured section model: collect following siblings until next header, put blocks in intro/wrapUp and hierElements as direct section children. Regenerate expected AKN for all IA test fixtures.

QR-TNA committed:
fix(leg): emit space instead of br in docTitle and inline for AKN 3.0

docTitle and inline in AKN 3.0 only allow b, i, u, a, ref, marker,
authorialNote, img—not br. Add AddInlineNoBr and use it when building
those elements so line breaks are output as a space and output validates.

Regenerate IA and EM test fixtures.

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
feat(leg): proprietary date/year/number|version and EM lookup by filename

Proprietary: IA use uk:date, uk:year, uk:number; EM use uk:date, uk:year, uk:version. EM mapping: key by CSV filename column, GetMappingRecord(filename); drop year/number/version parsing. Update test XSLT and expected AKN fixtures.

QR-TNA committed:
feat(leg): add uk:explanatoryMemoranda for EM, align IA to uk:impactAssessment

- EM: emit uk:explanatoryMemoranda with legislation.gov.uk document URI for contractor upload
- IA: fixtures now use uk:impactAssessment (was uk:IA)
- Regenerate all EM and IA .akn test fixtures

QR-TNA committed:
refactor(leg): use uk:associatedID for EM and IA upload URI

Replace uk:explanatoryMemoranda and uk:impactAssessment with single
uk:associatedID element for simpler validation and future document types.
Regenerate EM and IA .akn fixtures.

QR-TNA committed:
fix(em): use singular 'memorandum' in URIs instead of plural 'memoranda'

Change EM URI path segment from /memoranda/ to /memorandum/ to match
the correct singular form. Updated code generation logic and all 22 test
files to reflect the correct URI format.

QR-TNA committed:
feat(leg): extract images as GIF with S3 naming convention for EM and IA

- Add ConvertToGif() to src/img/Convert.cs using existing ImageSharp dependency
- Add LegImageProcessor to convert images to GIF and apply S3 naming convention
  ({type}_{year}{number}_{lang}_{seq:D3}.gif), traversing Header, Body and Annexes
- Add ImageNaming utility for generating S3-compatible filenames and paths
- Update TestEM and TestIA regeneration methods to save extracted GIF images
  alongside AKN test fixtures in the S3 folder structure
- Regenerate all EM and IA test AKN files with updated image src references
  (images/*.gif) and commit extracted GIF fixtures

QR-TNA committed:
fix(leg): prevent path traversal by sanitizing user input in path combination

Fixes insecure Path.Combine usage (CWE-22) by sanitizing user-supplied
input with Path.GetFileName before combining with directory paths.

QR-TNA committed:
feat(leg): add TOC link header to Impact Assessment HTML output

Renders a styled 'Impact Assessment Table of Contents' link at the top
of the page, built from request parameters, and suppresses the
in-document toc element.

QR-TNA committed:
Merge branch 'main' into LEG-45

QR-TNA committed:
feat(leg): add FRBRWork author for IA and EM associated documents

Populate FRBRauthor/@href in FRBRWork based on the legislation class of
the associated legislation, mapping to the appropriate legislature or
government body URI (e.g. government/uk, legislature/UnitedKingdomParliament).
Falls back to href="#" when the legislation class is unknown or absent.

QR-TNA committed:
fix(leg): propagate FRBRWork author to annex sub-documents

AnnexMetadata now copies WorkAuthor from the parent document metadata
so that annex FRBRWork sections carry the same legislature/government
author URI as the main document.

QR-TNA committed:
fix(leg): map UnitedKingdomImpactAssessment class to government/uk author

IAs with no linked legislation use UnitedKingdomImpactAssessment as their
legislation class. These are still UK government documents so their
FRBRWork author should be government/uk rather than falling back to "#".
